### PR TITLE
Docs enhancement

### PR DESCRIPTION
**Pull request by thuydotm**
This PR refactors code, update tests and examples for the following modules:
- aspect
- classify
- focal

Function `general_output_checks()` is added to test that output attributes are preserved.

### CODE REVIEW

File: xrspatial/aspect.py [hunk 1]
```diff
@@ -183,113 +196,90 @@ def aspect(agg: xr.DataArray,
...
+        >>> aspect_da = aspect(raster_da)
+        >>> print(aspect_da)
+        <xarray.DataArray 'aspect' (y: 6, x: 5)>
+        dask.array<_trim, shape=(6, 5), dtype=float64, chunksize=(3, 3), chunktype=numpy.ndarray>
+        Dimensions without coordinates: y, x
+        >>> aspect_da.compute()  # compute the results
+        <xarray.DataArray 'aspect' (y: 5, x: 5)>
+        array([[nan,     nan,        nan,        nan,          nan],
+               [nan, 120.96375653, 104.03624347, 135.        , nan],
+               [nan, 101.30993247, 108.43494882, 123.69006753, nan],
```
Comment (ianthomas23) [inline xrspatial/aspect.py:252]: I get values similar to the `numpy` values above. Maybe this is a copy and paste error?

File: xrspatial/tests/general_checks.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        # dask case
+        assert isinstance(
+            output_agg.data.compute(), type(input_agg.data.compute()))
+
+    # shape and other attributes remain the same
+    assert output_agg.shape == input_agg.shape
+    assert output_agg.dims == input_agg.dims
+    assert output_agg.attrs == input_agg.attrs
+    for coord in input_agg.coords:
+        assert np.all(output_agg[coord] == input_agg[coord])
```
Comment (ianthomas23) [inline xrspatial/tests/general_checks.py:20]: There is a `np.testing.assert_array_equal` that would be better here. But this applies throughout the codebase so maybe it should be in a separate issue/PR.

File: xrspatial/tests/general_checks.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    if expected_results is not None:
+        if isinstance(input_agg.data, da.Array):
+            assert np.isclose(
```
Comment (ianthomas23) [inline xrspatial/tests/general_checks.py:24]: Again for future reference `np.testing.assert_allclose` is a better alternative here.

File: xrspatial/utils.py [hunk 4]
```diff
@@ -98,6 +98,10 @@ def is_dask_cupy(agg: xr.DataArray):
     return isinstance(agg.data, da.Array) and is_cupy_backed(agg)
 
 
+def not_implemented_func(agg, *args):
+    raise NotImplementedError('Not yet implemented.')
```
Comment (ianthomas23) [inline xrspatial/utils.py:102]: The disadvantage here is that the exception is raised from a different point in the code than where the problem originally occurred. But the stack trace shows this, and the combined use of `not_implemented_func` and `ArrayTypeFunctionMapping` simplifies the code a lot so this is clearly a net improvement.

**Review by ianthomas23 — CHANGES_REQUESTED** [review]
A few comments, but only one thing to check and possibly change.

Thanks @thuydotm!

File: xrspatial/aspect.py (hunk 1 above)
Comment (thuydotm) [reply xrspatial/aspect.py]: Thanks for catching this, just fixed!

File: xrspatial/tests/general_checks.py (hunk 2 above)
Comment (thuydotm) [reply xrspatial/tests/general_checks.py:20]: sounds good, let's make this change in a separate PR

File: xrspatial/utils.py (hunk 4 above)
Comment (thuydotm) [reply xrspatial/utils.py:102]: We can add a `messages` param to the `not_implemented_func()` to clearly state the reason if it fails.

**Pull request merged**